cyclonedds/tools/cli/discovery/main.py

```diff
@@ -0,0 +1,243 @@
+                    break
+            else:
+                topic = DTopic(name=pub.endpoint.topic_name, subscriptions=[], publications=[pub])
+                par.topics.append(topic)
+
+        for sub in rdr.take(N=20, condition=rcr):
+            if sub.participant_key != dp.guid:
+                live.entities += 1
+
+            if not topic_re.match(sub.topic_name):
+                continue
+
+            sqos = sub.qos
+            naming = None
+            if core.Policy.EntityName in sqos:
+                naming = sqos[core.Policy.EntityName]
+                sqos = sqos - core.Qos(naming)
+
+            if sub.participant_key in participants:
+                par = participants[sub.participant_key]
+            else:
+                par = participants[sub.participant_key] = DParticipant(sample=None, topics=[])
+
+            sub = DPubSub(
+                endpoint=sub,
+                qos=sqos,
+                name=naming.name if naming else None
+            )
+
+            for topic in par.topics:
+                if topic.name == sub.endpoint.topic_name:
+                    topic.subscriptions.append(sub)
+                    break
+            else:
+                topic = DTopic(name=sub.endpoint.topic_name, subscriptions=[sub], publications=[])
+                par.topics.append(topic)
+
+        # yield thread
+        time.sleep(0.01)
+
+    live.result = list(participants.values())
+    live.delivered = True
+
+
+def ps_discovery(live: LiveData, domain_id: int, runtime: timedelta, show_self: bool, topic: str) -> List[PApplication]:
+    try:
+        topic_re = re.compile(f"^{topic}$")
+    except re.error:
+        topic_re = re.compile(f"^{re.escape(topic)}$")
+
+    dp = domain.DomainParticipant(domain_id)
+
+    rdp = builtin.BuiltinDataReader(dp, builtin.BuiltinTopicDcpsParticipant)
+    rcp = core.ReadCondition(rdp, core.SampleState.NotRead | core.ViewState.Any | core.InstanceState.Alive)
+    rdw = builtin.BuiltinDataReader(dp, builtin.BuiltinTopicDcpsPublication)
+    rcw = core.ReadCondition(rdw, core.SampleState.NotRead | core.ViewState.Any | core.InstanceState.Alive)
+    rdr = builtin.BuiltinDataReader(dp, builtin.BuiltinTopicDcpsSubscription)
+    rcr = core.ReadCondition(rdr, core.SampleState.NotRead | core.ViewState.Any | core.InstanceState.Alive)
+
+    applications = {}
+    participants = {}
+
+    hostname_get = core.Policy.Property("__Hostname", "")
+    appname_get = core.Policy.Property("__ProcessName", "")
+    pid_get = core.Policy.Property("__Pid", "")
+    address_get = core.Policy.Property("__NetworkAddresses", "")
+
+    start = datetime.now()
+    end = start + runtime
+    while datetime.now() < end and not live.terminate:
+        for p in rdp.take(N=20, condition=rcp):
+            if p.key == dp.guid and not show_self:
+                continue
+
+            hostname = p.qos[hostname_get].value if p.qos[hostname_get] is not None else "Unkown"
```

Comment:
"Unkown" is probably be a typo, and copied to the following three lines

Reply:
Good catch, fixed that.